README.md
# Action
The provided action is called "**sendAVR**".  

# Sensor
The provided sensor class is:
* AVRSensor

## Sensor attributes  
Possible attributes are:  
* volume
* display
* status

# Commands
### connect
Should only be used for debugging purposes. Creates a socket connection to the avr. Is automatically called on pimatic startup or before sending a data
command if necessary.

### disconnect
Should only be used for debugging purposes. Ends the connection to the avr, if exist.

### sendCommand
Has to be used in the following syntax (*value* is optional and only usable with "volume.set" at the moment):
```
[action] "[category].[command].[value]"
```
Sample:
*Set volume to -40dB (pioneer)*
```
sendAVR "volume.set.080"
```

*Switch the AVR off*
```
sendAVR "power.off"
```

## Supported commands
*handling the volume*
* volume.down
* volume.up
* volume.set
* volume.mute
* volume.status

*handling power status*
* power.on
* power.off
* power.status

*handling the input*
* input.phono
* input.cd
* input.cdr_tape
* input.dvd
* input.tv_sat
* input.video1
* input.video2
* input.dvr_bdr
* input.ipod_usb
* input.hdmi1
* input.hdmi2
* input.hdmi3
* input.hdmi4
* input.hdmi5
* input.hdmi6
* input.bd
* input.hmg
* input.internet_radio
* input.change (*cycle through available inputs*)
* input.status (*gets the currently selected input*)

*get currently shown display information*
* display.status

*change the listening mode*
* sound.extStereo
* sound.autoSurr
* sound.direct
* sound.pureDirect

*control the cursor*
* cursor.up
* cursor.down
* cursor.right
* cursor.left
* cursor.enter
* cursor.back

Additional commands for iPod control etc. could be found here but are not supported by this plugin yet. Feel free to contribute.  
Pioneer: http://www.pioneerelectronics.com/StaticFiles/PUSA/Files/Home%20Custom%20Install/VSX-1120-K-RS232.PDF  
Denon: http://openrb.com/wp-content/uploads/2012/02/AVR3312CI_AVR3312_PROTOCOL_V7.6.0.pdf

# Configuration
There are five (self explaining) configuration parameters
* host
* port
* maxVolume
* brand
* logLevel

### Sample Plugin Config:
```javascript    
    {
      "plugin": "pio-remote",
      "host": "192.168.0.15",
      "port": 23,
      "maxVolume": 100,
      "brand": "pioneer" || "denon" (default: pioneer),
      "logLevel": "info" || "debug" || "warn" || "error" (default: info)
    }
```
The available volume level depends on the AVR brand.  
**Pioneer**: Volume range from 000 (mute) till 185 (+12dB). Concrete value could be calculated by the following formula
```
dB = 0.5 * volLevel - 80,5
```
Sample for a limit of -50dB
```
-50 = 0.5 * volLevel - 80.5
30.5 = 0.5 * volLevel
61 = volLevel
```
**Denon**: Volume range from 000 (-80dB) till 098 (+18dB). Concrete value could be calculated by the following formula
```
dB = volLevel - 80
```
Sample for a limit of -50dB
```
-50 = volLevel - 80
30 = volLevel
```

### Sample Sensor Config:
```javascript    
    {
      "class": "AVRSensor",
      "id": "avrsensor",
      "name": "Display",
      "attributes": [
        {
          "name": "display"
        },
        {
          "name": "vol"
        },
        {
          "name": "status"
        }
      ],
      "xAttributeOptions": [
        {
          "name": "vol",
          "displaySparkline": false,
          "hidden": false
        },
        {
          "name": "status",
          "displaySparkline": false,
		  "hidden": true
        }
      ]
    }
```

### Sample Volume Slider config:  
*Slider*
```
    {
      "class": "DummyDimmer",
      "id": "dummy-dimmer",
      "name": "Slider"
    }
```    
*Rule*
```
    {
      "id": "change-vol",
      "name": "change vol",
      "rule": "if $dummy-dimmer.dimlevel changes then sendAvr \"volume.set.$dummy-dimmer.dimlevel\"",
      "active": true,
      "logging": true
    }
```


Port could be 23 or 8102 depending on the model.

# Known devices:
### Port 23:
* Pioneer SC-81
* Pioneer VSX-921
* Pioneer VSX-922

### Port 8102:
* Pioneer SC-81
* Pioneer VSX-42
* Pioneer VSX-43
* Pioneer VSX-527
* Pioneer VSX-528
* Pioneer VSX-921
* Pioneer VSX-1020
* Pioneer VSX-1023
* Pioneer VSX-2120
* Pioneer VSX-S510

# Beware
This plugin is in an early alpha stadium and you use it on your own risk. 
I'm not responsible for any possible damages that occur on your health, hard- or software.

# License
MIT


CHANGELOG.md
**version 0.5.3** *released 2017-03-19*
* Added retryCount increment and reset last retry after successful connect

**version 0.5.2** *released 2016-12-14*
* Added status variabe to receive the current power state of the device
* refactored interval timer

**version 0.5.1** *released 2016-10-03*
* Added clearTimeout call in destroy method

**version 0.5.0** *released 2016-10-03*
* Added destroy method to AVRSensor to make the plugin pimatic 0.9 compatible

**version 0.4.2** *released 2016-06-07*
* Added config parameter logLevel

**version 0.4.1** *released 2016-04-21*
* Added missing files to package.json

**version 0.4.0** *released 2016-04-18*
* Fix missing commands.json in npm package
* Outsourced response handler to make dedicated parsing for different brands possible

**version 0.3.1** *released 2015-12-18*
* fixed bug in min volume detection

**version 0.3.0** *released 2015-12-18*
* Added support for other AVR brands
* Adjusted volume slider logic to work for denon aswell
* Added denon sample values and reference to documentation
* Outsourced command config
* Added some sound commands for pioneer

**version 0.2.1** *released 2015-12-14*
* Fixed critical error that causes pimatic to crash if connection to AVR cannot be established

**version 0.2.0** *released 2015-12-12*
* AVR Volume is shown
* AVR Display is shown 1:1
* Volume Slider support
* maxVolume implemented



denonHandler.js
module.exports.handleData = function(stringyfiedData, env, cb) {          
  if(stringyfiedData.indexOf('MV') === 0) {
	var currentVol = stringyfiedData.substring(3,stringyfiedData.length -1) - 80;
  } else {
	env.logger.info('Received: ' + stringyfiedData);
  }
  
  var currentDisp;
  var currentState;
  var currentIn;
  
  cb(currentVol, currentDisp, currentState, currentIn);
};

module.exports.sendCommand = function (controlCodes, currentVolume, command, pluginConfig, client) {
  var brand = 'denon';

  /**
   * parse the command
   **/
  if (command && command.indexOf('\.') > 0) {
    var splittedCommand = command.split('\.');
    var category = splittedCommand[0];
    var func = splittedCommand[1];

    var maxVolume = pluginConfig.maxVolume || defaultMaxVolume;
    if (maxVolume > controlCodes[brand].maxVolume) {
      maxVolume = controlCodes[brand].maxVolume;
    }

    var volLevel = (currentVolume + 80.5) * 2;
    var value = '';

    /**
     * Handle max volume to avoid damage on user and equipment
     **/
    if (category === 'volume' && func === 'up') {
      if (volLevel >= maxVolume) {
        return 'Vol max reached!';
      }
    }

    if (category === 'volume' && func === 'down') {
      if (volLevel <= 1) {
        return 'Vol min reached!';
      }
    }

    if (splittedCommand.length === 3) {
      value = splittedCommand[2];
      /**
       * Calculate correct value depending on maxValue
       **/
      if (category === 'volume' && func === 'set') {
        if (category === 'volume' && value < 0) {
          value = '001';
        }

        value = Math.round(value * maxVolume / 100);
        while (value.toString().length < 2) {
          value = '0' + value;
        }
      }
    }

    if (controlCodes[brand][category][func]) {
      var dataToWrite = value + controlCodes[brand][category][func] + '\r';

      var dataSent = client.write(dataToWrite);
      if (dataSent) {
        return 'Sent command: ' + dataToWrite;
      } else {
        return 'An error occurred while sending data: ' + dataToWrite;
      }
    }
  } else {
    return 'Unsupported command: ' + command;
  }
};

pioneerHandler.js
module.exports.handleData = function (stringyfiedData, env, cb) {
  if (stringyfiedData.indexOf('VOL') === 0) {
    var currentVol = (0.5 * stringyfiedData.substring(3, 6) - 80.5);

  } else if (stringyfiedData.indexOf('FL0') === 0) {
    // handle default display
    var str = '';
    for (var i = 4; i < stringyfiedData.length; i += 2) {
      str += String.fromCharCode(parseInt(stringyfiedData.substr(i, 2), 16));
    }

    // Cut text to 15 characters
    if (str.length > 15) {
      str = str.substring(0, 15);
    }

    var currentDisp = str;
  } else if (stringyfiedData.indexOf('PWR') === 0) {	  
    // handle power state
	var tempState = stringyfiedData.substring(3, 4);
	
	// the states are not intuitive so we have to switch them
	if(tempState === '1') {
	  var currentState = 0;
	} else if (tempState === '0'){
	  var currentState = 1;
	} else {
	  env.logger.info('This power state is unknown: ' + tempState);
	}	
  } else if (stringyfiedData.indexOf('FN') === 0) {
    // handle input display
    var currentIn = stringyfiedData;
  } else {
    //env.logger.info('Received: ' + stringyfiedData);
  }
  cb(currentVol, currentDisp, currentState, currentIn);
};

module.exports.sendCommand = function (controlCodes, currentVolume, command, pluginConfig, client) {
  var brand = 'pioneer';

  /**
   * parse the command
   **/
  if (command && command.indexOf('\.') > 0) {
    var splittedCommand = command.split('\.');
    var category = splittedCommand[0];
    var func = splittedCommand[1];

    var maxVolume = pluginConfig.maxVolume || defaultMaxVolume;
    if (maxVolume > controlCodes[brand].maxVolume) {
      maxVolume = controlCodes[brand].maxVolume;
    }

    var volLevel = (currentVolume + 80.5) * 2;
    var value = '';

    /**
     * Handle max volume to avoid damage on user and equipment
     **/
    if (category === 'volume' && func === 'up') {
      if (volLevel >= maxVolume) {
        return 'Vol max reached!';
      }
    }

    if (category === 'volume' && func === 'down') {
      if (volLevel <= 1) {
        return 'Vol min reached!';
      }
    }

    if (splittedCommand.length === 3) {
      value = splittedCommand[2];
      /**
       * Calculate correct value depending on maxValue
       **/
      if (category === 'volume' && func === 'set') {
        if (category === 'volume' && value < 0) {
          value = '001';
        }

        value = Math.round(value * maxVolume / 100);
        while (value.toString().length < 3) {
          value = '0' + value;
        }
      }
    }

    if (controlCodes[brand][category][func]) {
      var dataToWrite = value + controlCodes[brand][category][func] + '\r';

      var dataSent = client.write(dataToWrite);
      if (dataSent) {
        return 'Sent command: ' + dataToWrite;
      } else {
        return 'An error occurred while sending data: ' + dataToWrite;
      }
    }
  } else {
    return 'Unsupported command: ' + command;
  }
};

pio-remote.js
var __bind = function (fn, me) {
    return function () {
      return fn.apply(me, arguments);
    };
  },
  __hasProp = {}.hasOwnProperty,
  __extends = function (child, parent) {
    for (var key in parent) {
      if (__hasProp.call(parent, key)) child[key] = parent[key];
    }
    function ctor() {
      this.constructor = child;
    }

    ctor.prototype = parent.prototype;
    child.prototype = new ctor();
    child.__super__ = parent.prototype;
    return child;
  };

module.exports = function (env) {
  var M, PioRemotePlugin, PioRemoteActionProvider, PioRemoteActionHandler, pioRemotePlugin, pioRemoteActionHandler, client, pluginConfig, connected;
  M = env.matcher;

  var Promise = env.require('bluebird');
  var net = env.require('net');
  var util = env.require('util');
  var assert = env.require('cassert');

  /**
   * Default connection parameters
   **/
  var defaultHost = '192.168.0.15';
  var defaultPort = 23;
  var defaultMaxVolume = 98;
  var defaultBrand = 'pioneer';

  var retryCount = 0;
  var lastRetry = new Date();
  var currentVolume = 0;
  var currentDisplay = '';
  var currentStatus = 0;
  var currentInput = '';
  var updateInterval = 2000;
  var avrHandler;
  var brand;
  var logLevel;


  /**
   * Additional codes (control iPod etc.) for pioneer or denon receivers could be found here:
   * http://www.pioneerelectronics.com/StaticFiles/PUSA/Files/Home%20Custom%20Install/VSX-1120-K-RS232.PDF
   * and here
   * http://openrb.com/wp-content/uploads/2012/02/AVR3312CI_AVR3312_PROTOCOL_V7.6.0.pdf
   **/
  var controlCodes = require("./commands.json");

  /**
   * THE PLUGIN ITSELF
   **/
  PioRemotePlugin = (function (_super) {
    __extends(PioRemotePlugin, _super);

    function PioRemotePlugin() {
      this.init = __bind(this.init, this);
      return PioRemotePlugin.__super__.constructor.apply(this, arguments);
    }

    PioRemotePlugin.prototype.init = function (app, framework, config) {
      var deviceConfigDef;
      this.framework = framework;
      this.config = config;
      pluginConfig = config;
      this.framework.ruleManager.addActionProvider(new PioRemoteActionProvider(this.framework));

      brand = pluginConfig.brand || defaultBrand;
      logLevel = pluginConfig.logLevel || "info";
	  
      avrHandler = require("./" + brand + "Handler.js");

      deviceConfigDef = require("./device-config-schema");

      this.framework.deviceManager.registerDeviceClass("AVRSensor", {
        configDef: deviceConfigDef.AVRSensor,
        createCallback: (function (_this) {
          return function (config) {
            return new AVRSensor(config, framework);
          };
        })(this)
      });
    };

    return PioRemotePlugin;

  })(env.plugins.Plugin);

  pioRemotePlugin = new PioRemotePlugin;
  /**
   * THE ACTION PROVIDER
   **/
  PioRemoteActionProvider = (function (_super) {
    __extends(PioRemoteActionProvider, _super);

    function PioRemoteActionProvider(framework, config) {
      this.framework = framework;
      this.config = config;
      this.connectAction = __bind(this.connectAction, this);
    }

    /**
     *This function handles action in the form of `sendAvr "some string"`
     **/

    PioRemoteActionProvider.prototype.parseAction = function (input, context) {
      var commandTokens, fullMatch, m, match, onEnd, retVal, setCommand;
      retVal = null;
      commandTokens = null;
      fullMatch = false;
      setCommand = (function (_this) {
        return function (m, tokens) {
          return commandTokens = tokens;
        };
      })(this);

      onEnd = (function (_this) {
        return function () {
          return fullMatch = true;
        };
      })(this);

      m = M(input, context).match("sendAvr ").matchStringWithVars(setCommand);

      if (m.hadMatch()) {
        match = m.getFullMatch();
        return {
          token: match,
          nextInput: input.substring(match.length),
          actionHandler: new PioRemoteActionHandler(this.framework, commandTokens)
        };
      } else {
        return null;
      }
    };

    return PioRemoteActionProvider;

  })(env.actions.ActionProvider);

  /**
   * THE ACTION HANDLER
   **/

  PioRemoteActionHandler = (function (_super) {
    __extends(PioRemoteActionHandler, _super);

    function PioRemoteActionHandler(framework, commandTokens) {
      this.framework = framework;
      this.commandTokens = commandTokens;
      this.executeAction = __bind(this.executeAction, this);
      this.connect = __bind(this.connect, this);
      this.disconnect = __bind(this.disconnect, this);
      this.sendCommand = __bind(this.sendCommand, this);
    }

    /**
     * Method for establishing connection to the receiver
     **/
    PioRemoteActionHandler.prototype.connect = function (command) {
      if (!client || !connected) {
        retryCount++;

        if(logLevel === "info") {
		  env.logger.info('Client isn\'t connected yet, establish new connection...');
		}
			
        client = new net.Socket();

        client.on('data', function (data) {
          pioRemoteActionHandler.handleData(data.toString());
        });

        /**
         * this event is triggered, if the connection is successfully established
         **/
        client.on('connect', function () {
		  if(logLevel === "info") {
			env.logger.info('Connection successfully established');
		  }
		  connected = true;
          retryCount = 0;
          lastRetry = new Date();
        });

        client.on('error', function (ex) {
		  if(logLevel === "info") {
            env.logger.info('An error occured during connecting to avr: ' + ex);
		  }	
          client.destroy();
          client = undefined;
          connected = false;
          
        });

        client.on('close', function () {
          if (client && connected) {
		    if(logLevel === "info") {
              env.logger.info('Connection closed');
            }
			client.disconnect();
          }
        });

        client.connect(pluginConfig.port ? pluginConfig.port : defaultPort,
          pluginConfig.host ? pluginConfig.host : defaultHost, function () {

            if (command) {
              /**
               * assuming everything went okay, because there is no callback
               **/
              return pioRemoteActionHandler.sendCommand(command);
            } else {
              return 'done';
            }
          });
      } else {
	    if(logLevel === "info") {
          env.logger.info('Client already connected, nothing to do.');
		}
      }

      return 'done';
    };

    /**
     * Method to call when connection isn't needed any longer
     **/
    PioRemoteActionHandler.prototype.disconnect = function () {
      if (client && connected) {
	    if(logLevel === "info") {
          env.logger.info('Closing telnet session...');
		}  
        client.destroy();
        client = undefined;
        connected = false;

        return 'disconnected';
      } else {
	    if(logLevel === "info") {
          env.logger.info('Nothing to disconnect from.');
		}  
      }

      return 'disconnected';
    };

    /**
     * Method for sending a command
     **/
    PioRemoteActionHandler.prototype.sendCommand = function (command) {
      return avrHandler.sendCommand(controlCodes, currentVolume, command, pluginConfig, client);
    };

    /**
     * Handle the incoming data
     **/
    PioRemoteActionHandler.prototype.handleData = function (stringyfiedData) {
      avrHandler.handleData(stringyfiedData, env, function (currentVol, currentDisp, currentState, currentIn) {
		if (currentVol) {
          currentVolume = currentVol;
        }
        if (currentDisp) {
          currentDisplay = currentDisp;
        }
		if (currentState === 0 || currentState === 1) {
          currentStatus = currentState;
        }
        if (currentIn) {
          currentInput = currentIn;
        }
      });
    };

    /**
     * This function handles action in the form of `sendAVR "command"`
     **/
    PioRemoteActionHandler.prototype.executeAction = function () {
      return this.framework.variableManager.evaluateStringExpression(this.commandTokens).then((function (_this) {
        return function (command) {
          switch (command) {
            case 'connect':
              return pioRemoteActionHandler.connect();
              break;
            case 'disconnect':
              return pioRemoteActionHandler.disconnect();
              break;
            default:
              /**
               * If no connection to the AVR exists, connect first and send the command afterwards
               **/
              if (!client || !connected) {
                return pioRemoteActionHandler.connect(command);
              } else {
                return pioRemoteActionHandler.sendCommand(command);
              }
          }
        };
      })(this));
    };

    return PioRemoteActionHandler;

  })(env.actions.ActionHandler);
  pioRemoteActionHandler = new PioRemoteActionHandler;

  /**
   * THE AVR SENSOR
   **/
  AVRSensor = (function (_super) {
    __extends(AVRSensor, _super);

    function AVRSensor(config, framework) {
      var attr;
      this.config = config;
      this.id = config.id;
      this.name = config.name;
      this.attributes = {};
	  this.intervalTimeoutObject = null;

      var func = (function (_this) {
        return function (attr) {
          var name = attr.name;

          assert(name === 'vol' || name === 'display' || name === 'status');

          switch (name) {
            case 'vol':
              _this.attributes[name] = {
                description: name,
                type: "number"
              };

              var getter = (function () {
                if (retryCount < 3 || (new Date - lastRetry) > 60000) {
                  if (!client) {
                    pioRemoteActionHandler.connect();
                  } else {
                    pioRemoteActionHandler.sendCommand('volume.status');
                  }
                }
                return Promise.resolve(currentVolume);
              });
              _this.attributes[name].unit = 'dB';
              _this.attributes[name].acronym = 'VOL';
              break;
            case 'display':
              _this.attributes[name] = {
                description: name,
                type: "string"
              };

              var getter = (function () {
                if (retryCount < 3 || (new Date - lastRetry) > 60000) {
                  if (!client) {
                    pioRemoteActionHandler.connect();
                  }
                }
                return Promise.resolve(currentDisplay);
              });
              break;
			case 'status':
              _this.attributes[name] = {
                description: name,
                type: "number"
              };

              var getter = (function () {
                if (retryCount < 3 || (new Date - lastRetry) > 60000) {
                  if (!client) {
                    pioRemoteActionHandler.connect();
                  }
                }
				return Promise.resolve(currentStatus);
              });
              break;  
            default:
              throw new Error("Illegal attribute name: " + name + " in AVRSensor.");
          }

          _this._createGetter(name, getter);		  
		  
          _this.intervalTimeoutObject = setInterval((function () {
            return getter().then(function (value) {
              return _this.emit(name, value);
            })["catch"](function (error) {
              return env.logger.debug(error.stack);
            });
          }), updateInterval);
		  return _this.intervalTimeoutObject;
		  
        };
      })(this);

      for (var i = 0; i < this.config.attributes.length; i++) {
        attr = this.config.attributes[i];
        func(attr);
      }

      AVRSensor.__super__.constructor.call(this);
    }

    return AVRSensor;

  })(env.devices.Sensor);

  AVRSensor.prototype.destroy = function() {
	if (this.intervalTimeoutObject != null) {
	  clearInterval(this.intervalTimeoutObject);
    }
    return AVRSensor.__super__.destroy.call(this);
  };
  
  module.exports.PioRemoteActionProvider = PioRemoteActionProvider;

  return pioRemotePlugin;
};